dot/parachain/candidate-validation/messages.go

```diff
@@ -31,14 +32,27 @@ type ValidateFromExhaustive struct {
 }
 
 // ValidationResult represents the result coming from the candidate validation subsystem.
+// Validation results can be either a ValidValidationResult or InvalidValidationResult.
+//
+// If the result is invalid,
+// store the reason for invalidity in the Err field of InvalidValidationResult.
+//
+// If the result is valid,
+// set the values of the CandidateCommitments and PersistedValidationData fields of ValidValidationResult.
 type ValidationResult struct {
-	IsValid                 bool
+	Value any
+}
+
+type ValidValidationResult struct {
 	CandidateCommitments    parachaintypes.CandidateCommitments
 	PersistedValidationData parachaintypes.PersistedValidationData
-	Err                     error
 }
 
-// PreCheck try to complie the given validation code and return the result
+type InvalidValidationResult struct {
+	ReasonForInvalidity error
+}
```

Comment:
Not like this. You did not make candidate invalidity error an enum. Any error could be passed in ReasonForInvalidity.

Below is how to restrict this using an error enum. `CandidateInvalidity` is an int, but also an error since it implements the error interface.


```
type CandidateInvalidity int 

const (
	InvalidOutputs    CandidateInvalidity = iota
	Timeout
	ParamsTooLarge
	CodeTooLarge
	//...
)

func (ic CandidateInvalidity) Error() string {
	switch ic {
	case InvalidOutputs:
		return "validation outputs check doesn't pass"
	case Timeout:
		return "execution timeout"
	case ParamsTooLarge:
		return "validation input is over the limit"
	case CodeTooLarge:
		return "code size is over the limit"
	//...
	}

	return "unknown invalidity reason"
}
```

Extend this code for all the errors mentioned in https://github.com/paritytech/polkadot-sdk/blob/686aa233e67c619bcdbc8b758a9ddf92c3315cf1/polkadot/node/primitives/src/lib.rs#L324

Reply:
I've updated this to make candidate invalidity an enum.